Improvements to OpenVarioMenu

Scumi opened the pull request:
Brief summary of the changes
----------------------------
Work-in-progress implementation update for the OpenVarioMenu.


Related issues and discussions
------------------------------
#1107 
#1207
Openvario/meta-openvario#199

Blaubart committed:
OV/OpenVarioMenu: set initial display rotation on startup

mk-it-easy committed:
OV/OpenVarioMenu: wait on startup to fix display rotation timing problem

MaxKellermann commented on src/OV/System.hpp:
```diff
@@ -0,0 +1,20 @@
+  TEMPORARY,
+};
+
+[[gnu::const]]
```
this attribute is wrong. Do you know the difference between "const" and "pure"?

Scumi replied:
Not sufficiently apparently. After educating myself a bit, I'm not sure if either fits here. I'd remove that one entirely.

MaxKellermann replied:
"pure" would be correct here - the return value can change if global state changes, but the return value will stay the same if you call it twice right way. "const" ignores global state.

MaxKellermann commented on src/OV/System.cpp:
```diff
@@ -3,10 +3,33 @@
+  if (value < 1) { value = 1; }
+  if (value > 10) { value = 10; }
+
+  File::WriteExisting(Path("/sys/class/backlight/lcd/brightness"), std::to_string(value).c_str());
```
No `std::to_string()` please - `fmt::format_int` is much faster and does not need to allocate anything.

Scumi replied:
Agreed, forgot about our new library.

MaxKellermann commented on src/OV/System.hpp:
```diff
@@ -9,6 +9,12 @@ enum class SSHStatus {
 };
 
+uint_least8_t
+OpenvarioGetBrightness();
```
noexcept?

Scumi replied:
added

MaxKellermann commented on src/OV/System.cpp:
```diff
@@ -4,10 +4,45 @@
+  BufferedOutputStream buffered(file);
+
+  for (const auto &i : map)
+    buffered.Fmt("{}={}\n", i.first.c_str(), i.second.c_str());
```
Why `c_str()`?

Scumi replied:
Not needed, removed.

MaxKellermann commented on src/OV/System.hpp:
```diff
@@ -3,12 +3,21 @@
 };
 
+void
+LoadConfigFile(ProfileMap &map, Path path);
```
Why use `ProfileMap` here? Is this a XCSoar profile? (There is no documentation explaining what this library is about)

Scumi replied:
`Profile/Map.hpp` is a nice library that I wanted to use to read and write `config.uEnv` - mainly for updating the `rotation` variable. However, Its probably too much for this rather small task and I can just use `std::map` and a few of its member functions for the intended job.

Documentation will be added.

Scumi commented on src/OV/System.cpp:
```diff
@@ -4,10 +4,45 @@
+    while (kvreader.Read(pair))
+      map.Set(pair.key, pair.value);
+  } catch (...) {
+    ShowError(std::current_exception(), "Config File");
```
Agreed, done.

Scumi commented:
> ```
> src/OV/SystemDialog.cpp:111:28: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
> ```

Did not observe this during make build with gcc, but could re-produce with clang. Fixed.

MaxKellermann commented:
More compiler complaints:
```
In file included from ./src/io/BufferedOutputStream.hxx:11,
                 from src/OV/System.cpp:10:
In constructor ‘fmt::v9::format_int::format_int(int)’,
    inlined from ‘void OpenvarioSetRotation(DisplayOrientation)’ at src/OV/System.cpp:112:41:
/usr/include/fmt/format.h:3674:54: error: ‘rotation’ may be used uninitialized [-Werror=maybe-uninitialized]
 3674 |   explicit format_int(int value) : str_(format_signed(value)) {}
      |                                         ~~~~~~~~~~~~~^~~~~~~
src/OV/System.cpp: In function ‘void OpenvarioSetRotation(DisplayOrientation)’:
src/OV/System.cpp:92:7: note: ‘rotation’ was declared here
   92 |   int rotation;
      |       ^~~~~~~~
```

MaxKellermann commented on src/OV/System.cpp:
```diff
@@ -113,6 +113,13 @@ OpenvarioSetRotation(DisplayOrientation orientation)
+void
+OpenvarioSetLanguage(const char* lang)
+{
+  Run("localectl", "set-locale", lang);
```
Did you test this? Does this work?

Scumi replied:
No, not tested yet. I should've marked that commit as WIP as well. Was just an idea as a first shot without further looking into it. I can keep such things out of this Draft PR if it wastes your time looking at it.

MaxKellermann commented on src/OV/System.cpp:
```diff
@@ -113,6 +113,13 @@ OpenvarioSetRotation(DisplayOrientation orientation)
+OpenvarioSetLanguage(const char* lang)
+{
+  Run("localectl", "set-locale", lang);
+  Run("export", fmt::format("LANG={}", lang).c_str());
```
This has zero effect, even if this function call were able to invoke a shell built-in (which it does not). So you spawn a shell, use `export` to set an environment variable in this process, and then the process exits. The environment change is gone.

Scumi commented:
> More compiler complaints:
> 
> [...]

May I ask what your compile command is?
I'm wondering why I do not see those compiler complaints.

Also you merged the rotation commit as is, right? I need to issue a fixup commit to have this variable initialized properly.

Scumi committed:
OV/System: swap LANDSCAPE and REVERSE_LANDSCAPE rotation values

The screen was rotated the wrong way after restart for both portrait orientations.

MaxKellermann commented on src/OV/OpenVarioMenu.cpp:
```diff
@@ -467,6 +468,9 @@ Main()
 
 int main()
 {
+  /*the x-menu is waiting a second to solve timing problem with display rotation */
+  std::this_thread::sleep_for(std::chrono::seconds(1));
```
This sounds really horrible, and the justification for such a horrible kludge must be documented really well. Just saying "timing problem" isn't enough.

lordfolken commented:
Cherry-picked the portrait rotation mapping fix onto a fresh branch (OV/System: align OpenvarioGetRotation/SetRotation with fbcon/DetectInitialOrientation). Left out the 1-second startup sleep and OpenVarioMenu InitialOrientation wiring — those can be a follow-up if needed. The rotation API was dead code on master; this makes it correct for future use.

Scumi commented:
The OpenVarioMenu will not be needed anymore once everything is integrated into XCSoar anyway.
